### PR TITLE
User sign up

### PR DESCRIPTION
**Pull request by ducminhvu**

### CODE REVIEW

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+      flash[:success] = I18n.t "welcome"
+      redirect_to root_url
+    else
+      render "new"
+    end
+  end
+
+  private
+    def user_params
+      params.require(:user).permit(:username, :email, :password,
```
Comment (phamkykhoi) [inline app/controllers/users_controller.rb]: Remove ( )

File: app/models/user.rb [hunk 2]
```diff
@@ -11,4 +11,7 @@ class User < ActiveRecord::Base
                     presence: true, length: {maximum: 255},
                     uniqueness: {case_sensitive: false}
   validates :username, presence: true, length: {maximum: 50}
+  validates :password, presence: true, length: {minimum: 6}
```
Comment (phamkykhoi) [inline app/models/user.rb]: Haven't password_confirmation

File: app/views/users/new.html.erb [hunk 3]
```diff
@@ -0,0 +1,23 @@
+<% provide :title, t("sign_up") %>
+<h1>Sign up</h1>
```
Comment (phamkykhoi) [inline app/views/users/new.html.erb]: I18n

File: app/views/users/new.html.erb [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for @user do |f| %>
+    <%= render "shared/error_messages" %>
```
Comment (phamkykhoi) [inline app/views/users/new.html.erb]: Send object to shared/error_messages.

File: app/controllers/users_controller.rb [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+    if @user.save
+      flash[:success] = I18n.t "welcome"
+      redirect_to root_url
+    else
+      render "new"
+    end
+  end
+
+  private
+    def user_params
```
Comment (phamkykhoi) [inline app/controllers/users_controller.rb]: Indent

File: app/controllers/users_controller.rb [hunk 6]
```diff
@@ -0,0 +1,21 @@
+class UsersController < ApplicationController
+  def new
+    @user = User.new
+  end
+
+  def create
+    @user = User.new user_params
+    if @user.save
+      flash[:success] = I18n.t "welcome"
```
Comment (phamkykhoi) [inline app/controllers/users_controller.rb]: Auto login after sign up

File: app/controllers/users_controller.rb [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+      redirect_to root_url
+    else
+      render "new"
+    end
+  end
+
+  private
+  def user_params
+    params.require(:user).permit :username, :email, :password,
+                                  :password_confirmation
```
Comment (phamkykhoi) [inline app/controllers/users_controller.rb]: Indent

File: app/views/users/new.html.erb [hunk 8]
```diff
@@ -0,0 +1,23 @@
+<% provide :title, t("sign_up") %>
+<h1><%= t("sign_up") %></h1>
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for @user do |f| %>
+    <%= render 'shared/error_messages', object: f.object %>
```
Comment (phamkykhoi) [inline app/views/users/new.html.erb]: '' => ""

File: Gemfile [hunk 9]
```diff
@@ -16,4 +17,3 @@ group :development, :test do
   gem "web-console", "~> 2.0"
   gem "spring"
 end
-
```
Comment (phamkykhoi) [inline Gemfile]: Check out file  Gemfile

File: Gemfile.lock [hunk 10]
```diff
@@ -148,6 +149,7 @@ PLATFORMS
   ruby
 
 DEPENDENCIES
+  bcrypt (= 3.1.7)
   bootstrap-sass (~> 3.3.5)
   byebug
   coffee-rails (~> 4.1.0)
```
Comment (phamkykhoi) [inline Gemfile.lock]: Check out this file

File: app/views/shared/_error_messages.html.erb [hunk 11]
```diff
@@ -0,0 +1,12 @@
+<% if object.errors.any? %>
+  <div id="error_explanation">
+    <div class="alert alert-danger">
+      The form contains <%= pluralize(object.errors.count, "error") %>.
+    </div>
+    <ul>
+    <% object.errors.full_messages.each do |msg| %>
```
Comment (phamkykhoi) [inline app/views/shared/_error_messages.html.erb]: Indent

File: app/models/user.rb [hunk 12]
```diff
@@ -3,12 +3,14 @@ class User < ActiveRecord::Base
...
   has_many :following, through: :active_relationships, source: :followed
 
   validates :email, format: {with: VALID_EMAIL_REGEX},
-                    presence: true, length: {maximum: 255},
-                    uniqueness: {case_sensitive: false}
+    presence: true, length: {maximum: 255}, uniqueness: {case_sensitive: false}
   validates :username, presence: true, length: {maximum: 50}
+  validates :password, presence: true, length: {minimum: 6}
+  
+  before_save ->{self.email = email.downcase}
```
Comment (phamkykhoi) [inline app/models/user.rb]: Move to after line VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i

File: app/views/shared/_error_messages.html.erb [hunk 13]
```diff
@@ -0,0 +1,12 @@
+<% if object.errors.any? %>
+  <div id="error_explanation">
+    <div class="alert alert-danger">
+      The form contains <%= pluralize object.errors.count, t("error") %>.
+    </div>
+    <ul>
+    <% object.errors.full_messages.each do |msg| %>
```
Comment (phamkykhoi) [inline app/views/shared/_error_messages.html.erb]: Indent

File: app/views/users/new.html.erb [hunk 14]
```diff
@@ -0,0 +1,23 @@
+<% provide :title, t("sign_up") %>
+<h1><%= t("sign_up") %></h1>
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for @user do |f| %>
+    <%= render 'shared/error_messages', :object => f.object %>
```
Comment (phamkykhoi) [inline app/views/users/new.html.erb]: ' ' => " "

Comment (phamkykhoi) [inline app/views/users/new.html.erb]: chuyển => thành  :

**Comment by phamkykhoi** [comment]
OK

File: app/controllers/users_controller.rb [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+      flash[:success] = I18n.t "welcome"
+      redirect_to root_url
+    else
+      render "new"
+    end
+  end
+
+  private
+  def user_params
+  params.require(:user).permit :username, :email, :password,
```
Comment (dieunb) [inline app/controllers/users_controller.rb]: indent

File: app/models/user.rb [hunk 16]
```diff
@@ -1,14 +1,16 @@
...
-                                  dependent: :destroy
+    foreign_key: "follower_id", dependent: :destroy
   has_many :following, through: :active_relationships, source: :followed
 
   validates :email, format: {with: VALID_EMAIL_REGEX},
-                    presence: true, length: {maximum: 255},
-                    uniqueness: {case_sensitive: false}
+    presence: true, length: {maximum: 255}, uniqueness: {case_sensitive: false}
   validates :username, presence: true, length: {maximum: 50}
+  validates :password, presence: true, length: {minimum: 6}
```
Comment (dieunb) [inline app/models/user.rb]: Store `6` in Setting

File: app/models/user.rb [hunk 17]
```diff
@@ -1,14 +1,16 @@
...
   has_many :active_relationships, class_name: "Relationship",
-                                  foreign_key: "follower_id",
-                                  dependent: :destroy
+    foreign_key: "follower_id", dependent: :destroy
   has_many :following, through: :active_relationships, source: :followed
 
   validates :email, format: {with: VALID_EMAIL_REGEX},
-                    presence: true, length: {maximum: 255},
-                    uniqueness: {case_sensitive: false}
+    presence: true, length: {maximum: 255}, uniqueness: {case_sensitive: false}
```
Comment (dieunb) [inline app/models/user.rb]: Store `255` in Setting

File: app/views/users/new.html.erb [hunk 18]
```diff
@@ -0,0 +1,23 @@
+<% provide :title, t("sign_up") %>
+<h1><%= t("sign_up") %></h1>
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for @user do |f| %>
```
Comment (dieunb) [inline app/views/users/new.html.erb]: slice form to new partial file

File: app/views/shared/_error_messages.html.erb [hunk 19]
```diff
@@ -0,0 +1,12 @@
+<% if object.errors.any? %>
+  <div id="error_explanation">
+    <div class="alert alert-danger">
+      The form contains <%= pluralize object.errors.count, t("error") %>.
```
Comment (khanhkhanh) [inline app/views/shared/_error_messages.html.erb]: i18n

File: app/views/static_pages/home.html.erb [hunk 20]
```diff
@@ -1,8 +1,8 @@
 <% provide :title, t("home") %>
 
-<div class="jumbotron">
+<div class="jumbotron center">
   <h1>Welcome to Framgia e-learning</h1>
   <h2>This is the home page for Framgia e-learning center.</h2>
 
-  <%= link_to t("sign_up"), "#", class: "btn btn-lg btn-primary" %>
+  <%= link_to t("sign_up"), signup_path, class: "btn btn-lg btn-primary" %>
```
Comment (khanhkhanh) [inline app/views/static_pages/home.html.erb]: Lines 4 + 5 => i18n

File: config/locales/en.yml [hunk 21]
```diff
@@ -7,4 +7,10 @@ en:
...
   log_in: "Log in"
-  sign_up: "Sign up now"
\ No newline at end of file
+  sign_up: "Sign up"
+  welcome: "Welcome to the Sample App!"
+  error: "error"
+  comfirmation: "Comfirmation"
+  create_my_account: "Create my account"
+  name: "Name"
+
```
Comment (khanhkhanh) [inline config/locales/en.yml]: Break line

File: app/views/users/new.html.erb [hunk 22]
```diff
@@ -0,0 +1,7 @@
+<% provide :title, t("sign_up") %>
+<h1><%= t("sign_up") %></h1>
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+  <%= render "layouts/signup_form"%>
```
Comment (phamkykhoi) [inline app/views/users/new.html.erb]: insert space  before %>

File: app/models/user.rb [hunk 23]
```diff
@@ -1,14 +1,18 @@
 class User < ActiveRecord::Base
   VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
-
+  before_save ->{self.email = email.downcase}
+  
   has_many :lessons, dependent: :destroy
   has_many :active_relationships, class_name: "Relationship",
-                                  foreign_key: "follower_id",
-                                  dependent: :destroy
+    foreign_key: "follower_id", dependent: :destroy
```
Comment (phamkykhoi) [inline app/models/user.rb]: Indent

File: app/models/user.rb [hunk 24]
```diff
@@ -1,14 +1,18 @@
 class User < ActiveRecord::Base
   VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
```
Comment (phamkykhoi) [inline app/models/user.rb]: Move Setttings

File: app/views/users/new.html.erb [hunk 25]
```diff
@@ -0,0 +1,7 @@
+<% provide :title, t("sign_up") %>
+<h1><%= t("sign_up") %></h1>
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+  <%= render "layouts/signup_form" %>
```
Comment (phamkykhoi) [inline app/views/users/new.html.erb]: Indent

File: app/models/user.rb [hunk 26]
```diff
@@ -1,14 +1,17 @@
 class User < ActiveRecord::Base
-  VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
-
+  before_save ->{self.email = email.downcase}
+
```
Comment (dieunb) [inline app/models/user.rb]: remove trailing spaces

**Comment by dieunb** [comment]
OK

**Pull request merged**